client/client-multi/client-c/src/kaa/platform-impl/common/common_key_utils.c

```diff
@@ -0,0 +1,391 @@
+}
+
+kaa_error_t ext_encrypt_data(uint8_t *input, size_t payload_size, uint8_t *output)
+{
+    if (!input || !payload_size || !output) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    /* Adding PKCS7 padding */
+    size_t enc_data_size;
+    ext_get_encrypted_data_size(payload_size, &enc_data_size);
+    uint8_t padding = enc_data_size - payload_size;
+    memset(input + payload_size, padding, padding);
+
+    /* Process the buffer chunk by chunk */
+    while (enc_data_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_ENCRYPT, input,
+                                KAA_SESSION_KEY_LENGTH,
+                                output, keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+
+        enc_data_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t ext_decrypt_data(uint8_t *input, size_t input_size,
+                             uint8_t *output, size_t *output_payload_size)
+{
+    if (!input || !input_size || !output || !output_payload_size) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    size_t tmp_in_size = input_size;
+    while (input_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_DECRYPT, input,
+                                KAA_SESSION_KEY_LENGTH, output,
+                                keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+        input_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    /* Reduce PKCS7 padding */
+    uint8_t padding_length = *(output - 1);
+    *output_payload_size = tmp_in_size - padding_length;
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t ext_get_signature(uint8_t *input, size_t input_size,
+                              uint8_t **output, size_t *output_size)
+{
+    if (!input || !input_size || !output || !output_size) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    kaa_error_t error = rsa_sign(&pk_context_, input, input_size, keys.signature, output_size);
+
+    if (error) {
+      return KAA_ERR_BADDATA;
+    }
+    *output = keys.signature;
+    *output_size = KAA_SIGNATURE_LENGTH;
+
+    return KAA_ERR_NONE;
+}
+
+/*
+ * Generates AES Key.
+ *
+ * bytes The size of the key.
+ * key   Generated AES Key.
+ *
+ */
+static int init_aes_key(unsigned char *key, size_t bytes)
+{
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_entropy_context entropy;
+
+    char *pers = "aes_generate_key";
+    int ret;
+
+    mbedtls_entropy_init(&entropy);
+    mbedtls_ctr_drbg_init(&ctr_drbg);
+
+    if ((ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy,
+                                     (unsigned char *) pers, strlen(pers))) != 0) {
+      printf(" failed\n ! mbedtls_ctr_drbg_init returned -0x%04x\n", -ret);
```

Comment:
Let's do not inject `kaa_context` here, since it is required for `KAA_LOG_xxx` macro. But leaving `printf` here IMO isn't good either. 
I propose to remove `printf` and just return indicative error code.